File: leetcode/[0543]diameter-of-binary-tree.ts
// https://leetcode.com/problems/diameter-of-binary-tree/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    return solution2(root);
}



// recursive post order: 1 ms
function solution1(root: TreeNode | null): number {
    let maxd = 0;

    const f = (n: TreeNode | null): number => {
        if (n) {
            const ld = f(n.left);
            const rd = f(n.right);
            const d = 1 + Math.max(ld, rd);
            maxd = Math.max(maxd, ld + rd);

            return d;
        } else {
            return 0;
        }
    };

    f(root);

    return maxd;
}



// iterative post order: 15 ms
function solution2(root: TreeNode | null): number {
    if (!root) return 0;

    const h = new WeakMap<TreeNode, number>();
    let maxd = 0;

    const st = [[root, false]] as [TreeNode, boolean][];
    while (st.length > 0) {
        const [n, v] = st.pop()!;
        if (n) {
            if (v) {
                const ld = h.get(n.left!) ?? 0;
                const rd = h.get(n.right!) ?? 0;
                const d = 1 + Math.max(ld, rd);
                maxd = Math.max(maxd, ld + rd);

                h.set(n, d);
            } else {
                st.push(
                    [n, true], [n.left!, false], [n.right!, false]
                );
            }
        }
    }

    return maxd;
}

File: leetcode/[0543]diameter-of-binary-tree.go
// https://leetcode.com/problems/diameter-of-binary-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	return solution2(root)
}



// recursive: 0 ms
func solution1(root *TreeNode) int {
	maxDiameter := 0

	var rec func(*TreeNode) int
	rec = func(n *TreeNode) int {
		if n == nil {
			return 0
		}

		lDepth, rDepth := rec(n.Left), rec(n.Right)
		maxDiameter = max(maxDiameter, lDepth+rDepth)
		return 1 + max(lDepth, rDepth)
	}
	rec(root)

	return maxDiameter
}



// iterative: 2 ms
func solution2(root *TreeNode) int {
	if root == nil {
		return 0
	}

	maxDiameter := 0
	ds := map[*TreeNode]int{}

	type S struct {
		Node        *TreeNode
		ToBeVisited bool
	}
	st := []S{S{root, false}}

	for len(st) > 0 {
		n := st[len(st)-1]
		st = st[:len(st)-1]

		if n.Node != nil {
			if n.ToBeVisited {
				lDepth, rDepth := ds[n.Node.Left], ds[n.Node.Right]
				maxDiameter = max(maxDiameter, lDepth+rDepth)
				ds[n.Node] = 1 + max(lDepth, rDepth)
			} else {
				st = append(
					st, S{n.Node, true}, S{n.Node.Right, false}, S{n.Node.Left, false},
				)
			}
		}
	}

	return maxDiameter
}
